### PR TITLE
draft pr check

### PR DESCRIPTION
**Pull request by Olowodarey**
pls help to check

### CODE REVIEW

File: contract/src/base/types.cairo [hunk 1]
```diff
@@ -144,5 +144,7 @@ pub struct PoolDetails {
     pub totalSharesOption2: u256,
     pub initial_share_price: u16,
     pub exists: bool,
+    pub assigned_validator1: starknet::ContractAddress,
```
Comment (Akshola00) [inline contract/src/base/types.cairo:147]: you dont neccesarily have to do this
you can use an array to store the two validators, 
you will still have to use a mapping of `map<pool_id, array<ContractAddress>>`  a pool to the array of validators

File: contract/src/interfaces/ipredifi.cairo [hunk 2]
```diff
@@ -49,4 +49,10 @@ pub trait IPredifi<TContractState> {
     fn manually_update_pool_state(
         ref self: TContractState, pool_id: u256, new_status: Status,
     ) -> Status;
+    fn get_pool_validators(
+        self: @TContractState, pool_id: u256,
+    ) -> (ContractAddress, ContractAddress);
```
Comment (Akshola00) [inline contract/src/interfaces/ipredifi.cairo:54]: good, but ca return an array instead

File: contract/src/predifi.cairo [hunk 3]
```diff
@@ -78,6 +78,8 @@ pub mod Predifi {
             u256, bool,
         >, // Pool ID to outcome (true = option2 won, false = option1 won)
         pool_resolved: Map<u256, bool>,
+        // Track which pools a validator is assigned to
+        validator_assignments: Map<(ContractAddress, u256), bool>,
```
Comment (Akshola00) [inline contract/src/predifi.cairo:82]: ive given the structure of the map above

File: contract/src/predifi.cairo [hunk 4]
```diff
@@ -160,11 +171,29 @@ pub mod Predifi {
...
+        // Get validators assigned to a specific pools
+        fn get_pool_validators(
+            self: @ContractState, pool_id: u256,
+        ) -> (ContractAddress, ContractAddress) {
+            let pool_details = self.pools.read(pool_id);
+            (pool_details.assigned_validator1, pool_details.assigned_validator2)
+        }
+
+
+        fn grant_validator_role(ref self: ContractState, validator: ContractAddress) {
```
Comment (Akshola00) [inline contract/src/predifi.cairo:188]: is this function necessary? All validators are in a vec in the state already

so you can get them from there

File: contract/src/predifi.cairo [hunk 5]
```diff
@@ -699,5 +733,64 @@ pub mod Predifi {
                 implied_probability2,
             }
         }
+
+
+        fn select_random_validators(ref self: ContractState) -> (ContractAddress, ContractAddress) {
```
Comment (Akshola00) [inline contract/src/predifi.cairo:738]: if you have written tests for this function and they have worked as expected then this is fine
this function fr now can only be called by an admin (if admin has not been implemented yet) then leave it as owner (it should only be called by owner)

but this function over all should randomly select from the vec of the contract addresses in the state and store the two addresses in the map declared above

File: contract/src/predifi.cairo [hunk 6]
```diff
@@ -699,5 +733,64 @@ pub mod Predifi {
...
+
+            // Update validator assignments
+            self.update_validator_assignments(validator1, self.pool_count.read());
+            self.update_validator_assignments(validator2, self.pool_count.read());
+
+            return (validator1, validator2);
+        }
+
+        /// Update validator assignments mapping
+        fn update_validator_assignments(
```
Comment (Akshola00) [inline contract/src/predifi.cairo:784]: no need for these, all can be done in a fucntion

File: contract/tests/test_contract.cairo [hunk 7]
```diff
@@ -1365,3 +1418,81 @@ fn test_validator_can_update_state() {
     assert(updated_pool.status == Status::Locked, 'should be updated by validator');
 }
 
+
+#[test]
+fn test_automated_validator_assignment() {
```
Comment (Akshola00) [inline contract/tests/test_contract.cairo:1423]: write detiaed test,
the first test should -
create a pool 
call the assign validator function with the owner as caller, and then assert that this function was called successfully
assert that the data in the pool struct was update and the validators have been added, you can know the contract addresses so no way to test for them there (since they are random) 

then assert that the map was updated and the pool now has two validators

File: contract/tests/test_contract.cairo [hunk 8]
```diff
@@ -1365,3 +1418,81 @@ fn test_validator_can_update_state() {
...
+    assigned_validator2 == _validator1 || 
+    assigned_validator2 == _validator2 || 
+    assigned_validator2 == _validator3 || 
+    assigned_validator2 == _validator4 ||
+    assigned_validator2 == contract_address_const::<0>(); // Zero address is valid for validator2
+
+assert(is_valid_validator1, 'Validator1 not from our list');
+assert(is_valid_validator2, 'Validator2 not valid');
+
+}
```
Comment (Akshola00) [inline contract/tests/test_contract.cairo:1498]: then add several tests taht test that the edge cases panic cases and lots more

**Review by Akshola00 — COMMENTED** [review]
nice work so far
there are some changes that should be made
